## Capstone Project (Exploratory Data Analysis)

Name: John Narvaez

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np

In [2]:
# Set a date parser
# dateParser = lambda[x: pd.datetime.strptime(x, '%Y-%m-%d %H:%M:%S')]

# Reading in the raw csv
# df = pd.read_csv('capstone/data/US_Accidents_March23.csv', index_col=0, parse_dates=['Start_Time', 'End_Time'], date_parser=dateParser)

In [3]:
# Reading in the raw csv
df = pd.read_csv('capstone/data/US_Accidents_March23.csv', index_col=0, parse_dates=['Start_Time', 'End_Time'])

In [4]:
df.head(3)

Source  Severity           Start_Time             End_Time  Start_Lat  \
ID                                                                            
A-1  Source2         3  2016-02-08 05:46:00  2016-02-08 11:00:00  39.865147   
A-2  Source2         2  2016-02-08 06:07:59  2016-02-08 06:37:59  39.928059   
A-3  Source2         2  2016-02-08 06:49:27  2016-02-08 07:19:27  39.063148   

     Start_Lng  End_Lat  End_Lng  Distance(mi)  \
ID                                               
A-1 -84.058723      NaN      NaN          0.01   
A-2 -82.831184      NaN      NaN          0.01   
A-3 -84.032608      NaN      NaN          0.01   

                                           Description  ... Roundabout  \
ID                                                      ...              
A-1  Right lane blocked due to accident on I-70 Eas...  ...      False   
A-2  Accident on Brice Rd at Tussing Rd. Expect del...  ...      False   
A-3  Accident on OH-32 State Route 32 Westbound at ...  ...      False   

    Station   Stop Traffic_Calming Traffic_Signal Turning_Loop Sunrise_Sunset  \
ID                                                                              
A-1   False  False           False          False        False          Night   
A-2   False  False           False          False        False          Night   
A-3   False  False           False           True        False          Night   

    Civil_Twilight Nautical_Twilight  Astronomical_Twilight  
ID                                                           
A-1          Night             Night                  Night  
A-2          Night             Night                    Day  
A-3          Night               Day                    Day  

[3 rows x 45 columns]

In [5]:
# Check the data frame info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7728394 entries, A-1 to A-7777761
Data columns (total 45 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Source                 object 
 1   Severity               int64  
 2   Start_Time             object 
 3   End_Time               object 
 4   Start_Lat              float64
 5   Start_Lng              float64
 6   End_Lat                float64
 7   End_Lng                float64
 8   Distance(mi)           float64
 9   Description            object 
 10  Street                 object 
 11  City                   object 
 12  County                 object 
 13  State                  object 
 14  Zipcode                object 
 15  Country                object 
 16  Timezone               object 
 17  Airport_Code           object 
 18  Weather_Timestamp      object 
 19  Temperature(F)         float64
 20  Wind_Chill(F)          float64
 21  Humidity(%)            float64
 22  Pressure(in)       

In [6]:
# Check the DataFrame how many rows and columns
rows, cols = df.shape
print(f'The DataFrame has rows - {rows:,} and columns - {cols}')

The DataFrame has rows - 7,728,394 and columns - 45


#### Dropping the columns

In [7]:
# List the columns
list(df)

['Source',
 'Severity',
 'Start_Time',
 'End_Time',
 'Start_Lat',
 'Start_Lng',
 'End_Lat',
 'End_Lng',
 'Distance(mi)',
 'Description',
 'Street',
 'City',
 'County',
 'State',
 'Zipcode',
 'Country',
 'Timezone',
 'Airport_Code',
 'Weather_Timestamp',
 'Temperature(F)',
 'Wind_Chill(F)',
 'Humidity(%)',
 'Pressure(in)',
 'Visibility(mi)',
 'Wind_Direction',
 'Wind_Speed(mph)',
 'Precipitation(in)',
 'Weather_Condition',
 'Amenity',
 'Bump',
 'Crossing',
 'Give_Way',
 'Junction',
 'No_Exit',
 'Railway',
 'Roundabout',
 'Station',
 'Stop',
 'Traffic_Calming',
 'Traffic_Signal',
 'Turning_Loop',
 'Sunrise_Sunset',
 'Civil_Twilight',
 'Nautical_Twilight',
 'Astronomical_Twilight']

In [8]:
# List of columns to drop
''' Use the list(df) or df.columns the created a variable "columns_to_drop" to hold the list
of columns to be dropped'''

# Assign a variable for the list of columns to be dropped (Copy & Paste)
columns_to_drop = ['Source', 'End_Lat',
 'End_Lng',
 'Distance(mi)',
 'Description', 'Zipcode',
 'Country',
 'Timezone',
 'Airport_Code',
 'Weather_Timestamp',
 'Temperature(F)',
 'Wind_Chill(F)',
 'Humidity(%)',
 'Pressure(in)',
 'Visibility(mi)',
 'Wind_Direction',
 'Wind_Speed(mph)',
 'Precipitation(in)',
 'Weather_Condition', 'Traffic_Calming',
 'Traffic_Signal',
 'Turning_Loop',
 'Sunrise_Sunset',
 'Civil_Twilight',
 'Nautical_Twilight',
 'Astronomical_Twilight' ]

In [9]:
# Dropping the columns listed in the 'columns_to_drop'
df = df.drop(columns=columns_to_drop)

In [10]:
# Re check if the desired columns are properly dropped.
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7728394 entries, A-1 to A-7777761
Data columns (total 19 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Severity    int64  
 1   Start_Time  object 
 2   End_Time    object 
 3   Start_Lat   float64
 4   Start_Lng   float64
 5   Street      object 
 6   City        object 
 7   County      object 
 8   State       object 
 9   Amenity     bool   
 10  Bump        bool   
 11  Crossing    bool   
 12  Give_Way    bool   
 13  Junction    bool   
 14  No_Exit     bool   
 15  Railway     bool   
 16  Roundabout  bool   
 17  Station     bool   
 18  Stop        bool   
dtypes: bool(10), float64(2), int64(1), object(6)
memory usage: 663.3+ MB


#### Converting to the correct data types

The data size originally was at 2+ GB and after dropping it was reduced to 663.3+ MB.
However, the Start_Time and End_Time has the object datatype.

In [11]:
# Single column conversion
# format = ''%Y-%M-%D %H:%M:%S' or 'ISO8601' or 'mixed'
# pd.to_datetime(df['Start_Time'], format='ISO8601')

'''Converting the Start_Time & End_Time column data type from object to datetime64[ns]'''
# Assign a variable to the slices of columns 'Start_Time to 'End_Time'
cols = df.columns[1:3]

# Apply a function doing a loop or iteration
df[cols] = df[cols].apply(pd.to_datetime, errors='coerce')

In [12]:
'''Convert the Street, City, County and State datatype from object to Category'''
# Assign a variable to the columns stated above
cols= df.columns[5:9]

# # Apply a function to convert to object to Category
df[cols] = df[cols].apply(pd.Categorical)

In [13]:
# Re check the data frame information and size
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7728394 entries, A-1 to A-7777761
Data columns (total 19 columns):
 #   Column      Dtype         
---  ------      -----         
 0   Severity    int64         
 1   Start_Time  datetime64[ns]
 2   End_Time    datetime64[ns]
 3   Start_Lat   float64       
 4   Start_Lng   float64       
 5   Street      category      
 6   City        category      
 7   County      category      
 8   State       category      
 9   Amenity     bool          
 10  Bump        bool          
 11  Crossing    bool          
 12  Give_Way    bool          
 13  Junction    bool          
 14  No_Exit     bool          
 15  Railway     bool          
 16  Roundabout  bool          
 17  Station     bool          
 18  Stop        bool          
dtypes: bool(10), category(4), datetime64[ns](2), float64(2), int64(1)
memory usage: 505.1+ MB


During initial view of the data the Severity is just one digit.
Below will check the values in the column to decide which data type is appropriate.

In [14]:
# Check the minimum and maximum values in the Severity column
Severity_min =df['Severity'].min()
Severity_max =df['Severity'].max()

print(f'The Severity min value is {Severity_min} and max values is {Severity_max}')

The Severity min value is 1 and max values is 4


In [15]:
'''
Since the values in the Severity column is just 1 digit.
Decided to convert it to int16. Did not convert to int8 just to give an allowance.
'''
# Convert the Severity data type from int64 to int16
df['Severity'] = df['Severity'].astype('int16')

In [16]:
# Re check the data frame info and size
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7728394 entries, A-1 to A-7777761
Data columns (total 19 columns):
 #   Column      Dtype         
---  ------      -----         
 0   Severity    int16         
 1   Start_Time  datetime64[ns]
 2   End_Time    datetime64[ns]
 3   Start_Lat   float64       
 4   Start_Lng   float64       
 5   Street      category      
 6   City        category      
 7   County      category      
 8   State       category      
 9   Amenity     bool          
 10  Bump        bool          
 11  Crossing    bool          
 12  Give_Way    bool          
 13  Junction    bool          
 14  No_Exit     bool          
 15  Railway     bool          
 16  Roundabout  bool          
 17  Station     bool          
 18  Stop        bool          
dtypes: bool(10), category(4), datetime64[ns](2), float64(2), int16(1)
memory usage: 460.9+ MB


Before down casting the Severity column the file size is 505+ MB.
After down casting it went down to 460+ MB.

#### Check null values

In [17]:
# Check the number of counts of empty data on each columns
df.isna().sum()

Severity           0
Start_Time    743166
End_Time      743166
Start_Lat          0
Start_Lng          0
Street         10869
City             253
County             0
State              0
Amenity            0
Bump               0
Crossing           0
Give_Way           0
Junction           0
No_Exit            0
Railway            0
Roundabout         0
Station            0
Stop               0
dtype: int64

Above output is not easy to read. Decided to make a output
that is easy to read and will compute the ratio of empty data rows of each column.

In [18]:
# Put the output of empty rows into variable col_dict
col = df.isna().sum()

# Convert the output to a dictionary
col_dict = col.to_dict()

# Get the maximum row count and put into a variable
num_of_rows = df.count().max()

# Print out the header/title
print('Missing Values count:')

# Looping into the list
for key, value in col_dict.items():
    # Check if row is not empty
    if value > 0:
        # Check the ratio or percentage agains the total of all row (empty and not empty)
        ratio = (value/num_of_rows)*100
        print(f'{key:<15}\t : {value:,} rows ({ratio:.2f} % of total)')

Missing Values count:
Start_Time     	 : 743,166 rows (9.62 % of total)
End_Time       	 : 743,166 rows (9.62 % of total)
Street         	 : 10,869 rows (0.14 % of total)
City           	 : 253 rows (0.00 % of total)


The Street and City columns empty data is just less than 1%. However, the 
Start_Time and End_Time is almost 10%. 

In [19]:
# Time delta between the start and last record date
df['Start_Time'].max() - df['Start_Time'].min()

Timedelta('2633 days 03:11:27')

#### Fill in the missing values in Start_Time and End_Time

In [20]:
'''Fill in the values of columns 'Start_Time' and 'End_Time using interpolation'''

# Passing the filled values by interpolation
df[['Start_Time', 'End_Time']] = df[['Start_Time', 'End_Time']].interpolate(method='linear')

# Check if there is no more null values
df.isna().sum()

Severity          0
Start_Time        0
End_Time          0
Start_Lat         0
Start_Lng         0
Street        10869
City            253
County            0
State             0
Amenity           0
Bump              0
Crossing          0
Give_Way          0
Junction          0
No_Exit           0
Railway           0
Roundabout        0
Station           0
Stop              0
dtype: int64

#### Dropping rows of empty data in columns 'Street' and 'City'

In [21]:
# Drop the cells of Street and City column that is NaN
df = df.dropna(subset=['Street', 'City'])

# Alternative using inplace
# df.dropna(subset=['Street', 'City'], inplace=True)

In [22]:
# Sanity check that all rows have no more missing values
df.isna().sum()

Severity      0
Start_Time    0
End_Time      0
Start_Lat     0
Start_Lng     0
Street        0
City          0
County        0
State         0
Amenity       0
Bump          0
Crossing      0
Give_Way      0
Junction      0
No_Exit       0
Railway       0
Roundabout    0
Station       0
Stop          0
dtype: int64

No more missing data in all columns and rows.

#### Duplicate rows


In [23]:
df.duplicated()

ID
A-1          False
A-2          False
A-3          False
A-4          False
A-5          False
             ...  
A-7777757    False
A-7777758    False
A-7777759    False
A-7777760    False
A-7777761    False
Length: 7717272, dtype: bool

It seems like no duplicated rows. Check using loc for any duplicated rows

In [24]:
# Check any duplicated rows. 2nd or more row that are similar
df.loc[df.duplicated()]

Severity          Start_Time            End_Time  Start_Lat  \
ID                                                                       
A-17911           2 2016-11-21 07:02:28 2016-11-21 07:32:02  38.501335   
A-27592           3 2016-08-17 09:14:48 2016-08-17 09:59:48  37.317421   
A-32098           3 2016-03-30 07:24:39 2016-03-30 08:24:39  37.751484   
A-35367           3 2016-04-20 17:33:58 2016-04-20 18:18:58  37.859638   
A-52481           2 2016-07-29 19:52:50 2016-07-29 20:37:50  33.687443   
...             ...                 ...                 ...        ...   
A-7776928         4 2019-08-22 18:01:59 2019-08-22 18:29:54  35.941291   
A-7777236         2 2019-08-22 17:17:36 2019-08-22 21:17:36  42.258280   
A-7777386         3 2019-08-23 17:59:11 2019-08-23 18:26:19  39.812370   
A-7777452         4 2019-08-23 12:14:35 2019-08-23 12:43:22  41.543690   
A-7777453         4 2019-08-23 12:14:35 2019-08-23 12:43:22  41.543690   

            Start_Lng           Street          City       County State  \
ID                                                                        
A-17911   -122.431770  Silverado Trl S  Saint Helena         Napa    CA   
A-27592   -121.935356   Sinclair Fwy N      San Jose  Santa Clara    CA   
A-32098   -121.470055          I-205 W         Tracy  San Joaquin    CA   
A-35367   -121.281227            I-5 S   French Camp  San Joaquin    CA   
A-52481   -117.872017          I-405 N     Santa Ana       Orange    CA   
...               ...              ...           ...          ...   ...   
A-7776928  -78.686140     Creedmoor Rd       Raleigh         Wake    NC   
A-7777236 -122.788930            I-5 N        Talent      Jackson    OR   
A-7777386  -74.985120     Gibbsboro Rd     Clementon       Camden    NJ   
A-7777452  -96.141010        N 19th St         Blair   Washington    NE   
A-7777453  -96.141010        N 19th St         Blair   Washington    NE   

           Amenity   Bump  Crossing  Give_Way  Junction  No_Exit  Railway  \
ID                                                                          
A-17911      False  False     False     False     False    False    False   
A-27592      False  False     False     False     False    False    False   
A-32098      False  False     False     False     False    False    False   
A-35367      False  False     False     False     False    False    False   
A-52481      False  False     False     False     False    False    False   
...            ...    ...       ...       ...       ...      ...      ...   
A-7776928    False  False     False     False     False    False    False   
A-7777236    False  False     False     False     False    False    False   
A-7777386    False  False     False     False     False    False    False   
A-7777452    False  False     False     False     False    False    False   
A-7777453    False  False     False     False     False    False    False   

           Roundabout  Station   Stop  
ID                                     
A-17911         False    False  False  
A-27592         False    False  False  
A-32098         False    False  False  
A-35367         False    False  False  
A-52481         False    False  False  
...               ...      ...    ...  
A-7776928       False    False  False  
A-7777236       False    False  False  
A-7777386       False    False  False  
A-7777452       False    False  False  
A-7777453       False    False  False  

[130119 rows x 19 columns]

It is quite surprising that it put out 372,192 rows and 19 columns duplicated.

In [25]:
# Pulling out a subset of columns to inspect
df.loc[df.duplicated(subset=['Start_Time', 'City', 'State', 'Bump', 'Amenity'])].head(3)

Severity          Start_Time            End_Time  Start_Lat  Start_Lng  \
ID                                                                              
A-105         2 2016-02-11 09:16:13 2016-02-11 10:01:13  39.769173 -84.207283   
A-612         3 2016-03-11 09:37:00 2016-03-11 11:37:00  39.950317 -83.017242   
A-704         2 2016-03-17 11:10:37 2016-03-17 13:29:20  39.783344 -84.253738   

            Street      City      County State  Amenity   Bump  Crossing  \
ID                                                                         
A-105  W Grand Ave    Dayton  Montgomery    OH    False  False     False   
A-612       I-70 E  Columbus    Franklin    OH    False  False     False   
A-704    Kings Hwy    Dayton  Montgomery    OH    False  False     False   

       Give_Way  Junction  No_Exit  Railway  Roundabout  Station   Stop  
ID                                                                       
A-105     False     False    False    False       False    False  False  
A-612     False     False    False    False       False    False  False  
A-704     False     False    False    False       False    False  False

Looking at the subset there seems no exactly duplicated rows. Conduct the last one specifying the 
datetime

In [26]:
# Check if the 'Start_Time column time has a duplicate'
df[(df['Start_Time']=='2016-02-11 09:16:13')]

# Alternative use query
# df.query('Start_Time'=='2016-02-11 09:16:13')

Severity          Start_Time            End_Time  Start_Lat  Start_Lng  \
ID                                                                              
A-104         2 2016-02-11 09:16:13 2016-02-11 10:01:13  39.691864 -84.211166   
A-105         2 2016-02-11 09:16:13 2016-02-11 10:01:13  39.769173 -84.207283   

            Street    City      County State  Amenity   Bump  Crossing  \
ID                                                                       
A-104     Lamme Rd  Dayton  Montgomery    OH    False  False     False   
A-105  W Grand Ave  Dayton  Montgomery    OH    False  False     False   

       Give_Way  Junction  No_Exit  Railway  Roundabout  Station   Stop  
ID                                                                       
A-104     False     False    False    False       False    False  False  
A-105     False     False    False    False       False    False  False

The above shows the **'Start_Time'**<u> is the same</u> but looking at the **'Street', 'Start_Lat' and 'Start_Lng'**
thus <u>there is no exactly rows with the same data in all columns.</u>

Thus at this time I assume there is no duplicates.

#### Renaming columns

In [27]:
# Rename the columns 'Start_Lat' to 'Latitude' and 'Start_Lng' to 'Longitude'
df = df.rename(columns={'Start_Lat': 'Latitude', 'Start_Lng' : 'Longitude'})

# Re-check if the column name is properly renamed
df.head(3)

Severity          Start_Time            End_Time   Latitude  Longitude  \
ID                                                                            
A-1         3 2016-02-08 05:46:00 2016-02-08 11:00:00  39.865147 -84.058723   
A-2         2 2016-02-08 06:07:59 2016-02-08 06:37:59  39.928059 -82.831184   
A-3         2 2016-02-08 06:49:27 2016-02-08 07:19:27  39.063148 -84.032608   

             Street          City      County State  Amenity   Bump  Crossing  \
ID                                                                              
A-1          I-70 E        Dayton  Montgomery    OH    False  False     False   
A-2        Brice Rd  Reynoldsburg    Franklin    OH    False  False     False   
A-3  State Route 32  Williamsburg    Clermont    OH    False  False     False   

     Give_Way  Junction  No_Exit  Railway  Roundabout  Station   Stop  
ID                                                                     
A-1     False     False    False    False       False    False  False  
A-2     False     False    False    False       False    False  False  
A-3     False     False    False    False       False    False  False

#### Saving the cleaned data into a csv

In [28]:
# Save the cleaned csv
df.to_csv('capstone/data/US_Accidents_Cleaned.csv')

# df.to_csv('capstone/data/US_Accidents_Cleaned.csv', index=False)

In [29]:
# Check if the cleaned csv is saved in the capstone/data folder
# import pandas as pd
# pd.read_csv('capstone/data/US_Accidents_Cleaned.csv')